<h1 style="font-size:3rem;color:rgb(0, 91, 94);text-align:center;">Topic 2 - Randomness</h1>
<hr style=\"border-top: 1px solid rgb(0, 91, 94);\" />

<h2 style="color:rgb(0, 91, 94)">Exercise 1</h2>
<hr style=\"border-top: 1px solid rgb(0, 91, 94);\" />

It is somewhat interesting that (5 * 4 * 3 * 2 * 1) perfectly divides (10 * 9 * 8 * 7 * 6) - there's no remainder.

If we only wanted exactly four heads as opposed to five, the equivalent calculation would be (10 * 9 * 8 * 7) / (4 * 3 * 2 * 1).

Does that evenly divide too? What is the formula in general?

Does it always come out as a positive whole number?

In [7]:
(10*9*8*7) / (4*3*2*1)

210.0

Does (10 * 9 * 8 * 7) / (4 * 3 * 2 * 1) divide evenly? 

Yes; (10*9*8*7) / (4*3*2*1) = 210

What is the formula in general? 

The formula in general is:
Divide the result of multiplying the all possible positions of heads (based on the number of heads you want) in a single 10 coin flip by the result of multiplying the divisions that will need to be done to account for combinations counted more than once.

For example if you wanted to count the unique positions of an outcome of just 2 head from the 10 coin flips:

(10*9) / (2*1)
10, to account for the first position of the heads.
9, to account for the second position of the heads.
2, to account for the combinations counted twice.
1, to make the example look more appealing.

Does it always come out as a positive whole number?

Yes, provided there is at least 1 heads in a set of 10 coin flips.

*note for 10 heads there is only 1 permutation.

In [12]:
count = 9
nine_heads = (10*9*8*7*6*5*4*3*2)/(9*8*7*6*5*4*3*2*1)
eight_heads = (10*9*8*7*6*5*4*3)/(8*7*6*5*4*3*2*1)
seven_heads = (10*9*8*7*6*5*4)/(7*6*5*4*3*2*1)
six_heads = (10*9*8*7*6*5)/(6*5*4*3*2*1)
five_heads = (10*9*8*7*6)/(5*4*3*2*1)
four_heads = (10*9*8*7)/(4*3*2*1)
three_heads = (10*9*8)/(3*2*1)
two_heads = (10*9)/(2*1)
one_heads = 10.0

calculations = [nine_heads, eight_heads, seven_heads, six_heads, five_heads, four_heads, three_heads, two_heads, one_heads]

for i in calculations:
    print(f'The number of unique positions for {count} heads = {i}')
    count += -1

The number of unique positions for 9 heads = 10.0
The number of unique positions for 8 heads = 45.0
The number of unique positions for 7 heads = 120.0
The number of unique positions for 6 heads = 210.0
The number of unique positions for 5 heads = 252.0
The number of unique positions for 4 heads = 210.0
The number of unique positions for 3 heads = 120.0
The number of unique positions for 2 heads = 45.0
The number of unique positions for 1 heads = 10.0


<h2 style="color:rgb(0, 91, 94)">Exercise 2</h2>
<hr style=\"border-top: 1px solid rgb(0, 91, 94);\" />

Note that there are the same number of ways to get 4 tails as there to get 4 heads. Explain why this is.

<h2 style="color:rgb(0, 91, 94)">Exercise 3</h2>
<hr style=\"border-top: 1px solid rgb(0, 91, 94);\" />

Plot bar charts or histograms of any three different distributions listed at the following link:

https://numpy.org/doc/stable/reference/random/generator.html#distributions